WcaOnRails/app/models/result.rb

```diff
@@ -1,15 +1,86 @@
 # frozen_string_literal: true
 class Result < ActiveRecord::Base
+  include ResultMethods
+
   self.table_name = "Results"
 
   belongs_to :competition, foreign_key: :competitionId
+  belongs_to :country, foreign_key: :countryId
   belongs_to :person, -> { current }, primary_key: :wca_id, foreign_key: :personId
   belongs_to :round, foreign_key: :roundId
   belongs_to :event, foreign_key: :eventId
+  belongs_to :format, foreign_key: :formatId
 
   scope :podium, -> { joins(:round).merge(Round.final_rounds).where(pos: [1..3]).where("best > 0") }
   scope :winners, -> { joins(:round, :event).merge(Round.final_rounds).where("pos = 1 and best > 0").order("Events.rank") }
 
+  validate :best_and_average
+  def best_and_average
+    return errors.add(:competitionId, "invalid") unless competition
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
+
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+
+    if solve_times.all?(&:skipped?)
+      errors.add(:base, "Cannot skip all solves.")
+      return
```

Comment:
Aha. That makes sense!

I don't think it applies to all the returns, like the `nil` checks in line 19-23. Some of the others don't look forced either, but I don't know enough to say for sure.

The way to get around having returns all over would be making them an array of Lambdas that you execute until you've found an error. Or you can set booleans instead of returning. `average_is_computable = false`, and then wrap the rest of the code in `if average_is_computable`, etc. Both strike me as a cure worse than the disease.

Anyway. I think it would help a lot with readability to have a short comment explaining each exit. If someone needs to add a new validation, it would be pretty hard to understand where in the hierarchy to insert it otherwise.